File: src/main/java/ludwig/utils/PrettyPrinter.java
package ludwig.utils;

import ludwig.model.*;

import java.util.Comparator;

public class PrettyPrinter implements NodeVisitor<Void> {
    private final StringBuilder builder = new StringBuilder();

    private PrettyPrinter() {
    }

    public static String print(Node parent) {
        PrettyPrinter printer = new PrettyPrinter();
        parent.accept(printer);
        return printer.toString();
    }

    @Override
    public String toString() {
        if (builder.length() == 0 || builder.charAt(builder.length() - 1) != '\n') {
            builder.append('\n');
        }
        return builder.toString();
    }

    private int indentation;

    @Override
    public Void visitList(ListNode node) {
        print("list");
        boolean inline = level(node) < 4;
        for (Node n : node.children()) {
            child(n, inline);
        }
        return null;
    }

    @Override
    public Void visitFunction(FunctionNode functionNode) {
        boolean body = false;
        for (Node child : functionNode.children()) {
            if (!(child instanceof VariableNode)) {
                body = true;
            }
            if (body) {
                child(child, false);
            }
        }
        return null;
    }


    @Override
    public Void visitClass(ClassNode classNode) {
        return null;
    }

    @Override
    public Void visitCatch(CatchNode catchNode) {
        print("catch");
        catchNode.children().forEach(node -> child(node, false));
        return null;
    }


    @Override
    public Void visitLiteral(LiteralNode literalNode) {
        print(literalNode.text());
        return null;
    }

    @Override
    public Void visitPackage(PackageNode packageNode) {
        return null;
    }

    @Override
    public Void visitVariable(VariableNode variableNode) {
        print(variableNode.name());
        return null;
    }

    @Override
    public Void visitReference(ReferenceNode referenceNode) {
        print(referenceNode.toString());
        boolean inline = canBeInlined(referenceNode);
        referenceNode.children().forEach(node -> child(node, inline));
        return null;
    }

    private static boolean canBeInlined(Node<?> node) {
        if (level(node) > 3) {
            return false;
        }

        for (int i = 0; i < node.children().size() - 1; i++) {
            if (NodeUtils.INSTANCE.argumentsCount(node.children().get(i)) > 3) {
                return false;
            }
        }

        return true;
    }

    @Override
    public Void visitFunctionReference(FunctionReferenceNode functionReference) {
        print("ref ");
        if (!functionReference.children().isEmpty()) {
            print(functionReference.children().get(0).toString());
        }
        return null;
    }

    @Override
    public Void visitThrow(ThrowNode throwNode) {
        print("throw ");
        throwNode.children().get(0).accept(this);
        return null;
    }

    @Override
    public Void visitTry(TryNode tryNode) {
        print("try");
        tryNode.children().forEach(node -> child(node, false));
        return null;
    }

    @Override
    public Void visitPlaceholder(PlaceholderNode placeholderNode) {
        print(placeholderNode.toString());
        return null;
    }

    @Override
    public Void visitBreak(BreakNode breakNode) {
        print("break ");
        breakNode.children().get(0).accept(this);
        return null;
    }

    @Override
    public Void visitContinue(ContinueNode continueNode) {
        print("continue ");
        continueNode.children().get(0).accept(this);
        return null;
    }

    @Override
    public Void visitOverride(OverrideNode overrideNode) {
        for (int i = 1; i < overrideNode.children().size(); i++) {
            child(overrideNode.children().get(i), false);
        }
        return null;
    }

    @Override
    public Void visitCall(CallNode callNode) {
        print("call");
        child(callNode.children().get(0), true);
        boolean inline = level(callNode) < 4;
        for (int i = 1; i < callNode.children().size(); i++) {
            child(callNode.children().get(i), inline);
        }

        return null;
    }

    @Override
    public Void visitLambda(LambdaNode lambdaNode) {
        print("λ");

        boolean body = false;
        boolean inline = level(lambdaNode) < 4;
        for (Node n : lambdaNode.children()) {
            if (!body && !(n instanceof VariableNode)) {
                body = true;
                print(" : ");
            }
            if (body) {
                child(n, inline);
            } else {
                print(" " + n);
            }
        }
        return null;
    }

    @Override
    public Void visitReturn(ReturnNode returnNode) {
        print("return");
        returnNode.children().forEach(node -> child(node, true));
        return null;
    }

    @Override
    public Void visitProject(ProjectNode projectNode) {
        return null;
    }

    @Override
    public Void visitIf(IfNode ifNode) {
        print("if");
        child(ifNode.children().get(0), true);
        for (int i = 1; i < ifNode.children().size(); i++) {
            child(ifNode.children().get(i), false);
        }
        return null;
    }

    @Override
    public Void visitAssignment(AssignmentNode assignmentNode) {
        print("= ");
        child(assignmentNode.children().get(0), true);
        boolean inline = level(assignmentNode) < 4;
        for (int i = 1; i < assignmentNode.children().size(); i++) {
            child(assignmentNode.children().get(i), inline);
        }
        return null;
    }

    @Override
    public Void visitElse(ElseNode elseNode) {
        print("else");
        elseNode.children().forEach(node -> child(node, false));
        return null;
    }

    @Override
    public Void visitFor(ForNode forNode) {
        print("for");
        for (int i = 0; i < forNode.children().size(); i++) {
            child(forNode.children().get(i), i < 2);
        }
        return null;
    }

    private void child(Node node, boolean inline) {
        indentation++;
        if (!inline) {
            if (builder.length() > 0) {
                builder.append('\n');
            }
            for (int i = 1; i < indentation; i++) {
                print("    ");
            }
        } else {
            print(" ");
        }
        node.accept(this);
        indentation--;
    }

    void print(String s) {
        builder.append(s);
    }

    private static int level(Node<?> node) {
        return node.children().stream().map(PrettyPrinter::level).max(Comparator.naturalOrder()).orElse(0) + 1;
    }

}

File: src/main/kotlin/ludwig/utils/PrettyPrinter.kt
package ludwig.utils

import ludwig.model.*
import java.util.*

class PrettyPrinter private constructor() : NodeVisitor<Unit> {
    private val builder = StringBuilder()

    override fun toString(): String {
        if (builder.length == 0 || builder[builder.length - 1] != '\n') {
            builder.append('\n')
        }
        return builder.toString()
    }

    private var indentation: Int = 0

    override fun visitList(node: ListNode) {
        print("list")
        val inline = level(node) < 4
        for (n in node) {
            child(n, inline)
        }
    }

    override fun visitFunction(functionNode: FunctionNode) {
        var body = false
        for (child in functionNode) {
            if (child !is VariableNode) {
                body = true
            }
            if (body) {
                child(child, false)
            }
        }
    }


    override fun visitClass(classNode: ClassNode) {
    }

    override fun visitCatch(catchNode: CatchNode) {
        print("catch")
        catchNode.forEach { node -> child(node, false) }
    }


    override fun visitLiteral(literalNode: LiteralNode) {
        print(literalNode.text)
    }

    override fun visitPackage(packageNode: PackageNode) {
    }

    override fun visitVariable(variableNode: VariableNode) {
        print(variableNode.name)
    }

    override fun visitReference(referenceNode: ReferenceNode) {
        print(referenceNode.toString())
        val inline = canBeInlined(referenceNode)
        referenceNode.forEach { node -> child(node, inline) }
    }

    override fun visitFunctionReference(functionReference: FunctionReferenceNode) {
        print("ref ")
        if (!functionReference.isEmpty()) {
            print(functionReference[0].toString())
        }
    }

    override fun visitThrow(throwNode: ThrowNode) {
        print("throw ")
        throwNode[0].accept(this)
    }

    override fun visitTry(tryNode: TryNode) {
        print("try")
        tryNode.forEach { node -> child(node, false) }
    }

    override fun visitPlaceholder(placeholderNode: PlaceholderNode) {
        print(placeholderNode.toString())
    }

    override fun visitBreak(breakNode: BreakNode) {
        print("break ")
        breakNode[0].accept(this)
    }

    override fun visitContinue(continueNode: ContinueNode) {
        print("continue ")
        continueNode[0].accept(this)
    }

    override fun visitOverride(overrideNode: OverrideNode) {
        for (i in 1..overrideNode.size - 1) {
            child(overrideNode[i], false)
        }
    }

    override fun visitCall(callNode: CallNode) {
        print("call")
        child(callNode[0], true)
        val inline = level(callNode) < 4
        for (i in 1..callNode.size - 1) {
            child(callNode[i], inline)
        }
    }

    override fun visitLambda(lambdaNode: LambdaNode) {
        print("λ")

        var body = false
        val inline = level(lambdaNode) < 4
        for (n in lambdaNode) {
            if (!body && n !is VariableNode) {
                body = true
                print(" : ")
            }
            if (body) {
                child(n, inline)
            } else {
                print(" " + n)
            }
        }
    }

    override fun visitReturn(returnNode: ReturnNode) {
        print("return")
        returnNode.forEach { node -> child(node, true) }
    }

    override fun visitProject(projectNode: ProjectNode) {
    }

    override fun visitIf(ifNode: IfNode) {
        print("if")
        child(ifNode[0], true)
        for (i in 1 until ifNode.size) {
            child(ifNode[i], false)
        }
    }

    override fun visitAssignment(assignmentNode: AssignmentNode) {
        print("= ")
        child(assignmentNode[0], true)
        val inline = level(assignmentNode) < 4
        for (i in 1..assignmentNode.size - 1) {
            child(assignmentNode[i], inline)
        }
    }

    override fun visitElse(elseNode: ElseNode) {
        print("else")
        elseNode.forEach { node -> child(node, false) }
    }

    override fun visitFor(forNode: ForNode) {
        print("for")
        for (i in 0..forNode.size - 1) {
            child(forNode[i], i < 2)
        }
    }

    private fun child(node: Node, inline: Boolean) {
        indentation++
        if (!inline) {
            if (builder.length > 0) {
                builder.append('\n')
            }
            for (i in 1..indentation - 1) {
                print("    ")
            }
        } else {
            print(" ")
        }
        node.accept(this)
        indentation--
    }

    internal fun print(s: String?) {
        builder.append(s)
    }

    companion object {

        fun print(parent: Node): String {
            val printer = PrettyPrinter()
            parent.accept(printer)
            return printer.toString()
        }

        private fun canBeInlined(node: Node): Boolean {
            if (level(node) > 3) {
                return false
            }

            for (i in 0..node.size - 1 - 1) {
                if (NodeUtils.argumentsCount(node[i]) > 3) {
                    return false
                }
            }

            return true
        }

        private fun level(node: Node): Int {
            return node.stream().map<Int>({ level(it) }).max(Comparator.naturalOrder()).orElse(0) + 1
        }
    }

}
